### PR TITLE
[SPARK-47976][Python] Add Python UDTF DataFrame API to return result of 'analyze' method as output table

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR adds a Python UDTF DataFrame API to return the result of calling the 'analyze' method as the output table, instead of the 'eval' and 'terminate' methods.

For example:

```
val udtfRunAnalyzeMethod: UserDefinedPythonTableFunction =
  udtf.copy(returnResultOfAnalyzeMethod = true)
val df = udtfRunAnalyzeMethod(spark, lit("abc"))
  .select("schema")
  .collect()
```

This should return the 'analyze' method result schema, with something like:

```
{
  "schema": {
    "fields": [
      {
        "metadata": {},
        "name": "partition_col",
        "nullable": true,
        "type": "integer"
      },
      {
        "metadata": {},
        "name": "count",
        "nullable": true,
        "type": "integer"
      },
      {
        "metadata": {},
        "name": "last",
        "nullable": true,
        "type": "integer"
      }
    ],
    "type": "struct"
  },
  "withSinglePartition": "False",
  "partitionBy": [
    {
      "name": "partition_col"
    }
  ],
  "orderBy": [
    {
      "name": "input",
      "ascending": "True",
      "overrideNullsFirst": "None"
    }
  ],
  "select": [
    {
      "name": "partition_col",
      "alias": ""
    },
    {
      "name": "input",
      "alias": ""
    }
  ]
}
```

### Why are the changes needed?

This can be useful for moving this computation from drivers to executors if the Python is expensive.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This PR adds SQL query golden file test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @allisonwang-db @ueshin 👀 ❓

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala [hunk 1]
```diff
@@ -252,17 +261,24 @@ case class UnresolvedPolymorphicPythonUDTF(
  *                                 projection to evaluate these expressions and return the result to
  *                                 the UDTF. The UDTF then receives one input column for each
  *                                 expression in the list, in the order they are listed.
- * @param pickledAnalyzeResult this is the pickled 'AnalyzeResult' instance from the UDTF, which
- *                             contains all metadata returned by the Python UDTF 'analyze' method
- *                             including the result schema of the function call as well as optional
- *                             other information
+ * @param partitionByStrings String representations of the partitioning expressions before parsing.
+ * @param orderByStrings String representations of the ordering expressions before parsing.
+ * @param selectedInputStrings String representations of the selected input expressions before
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:266]: why do we need the string representation here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/pythonLogicalOperators.scala [hunk 2]
```diff
@@ -274,6 +274,24 @@ case class ArrowEvalPythonUDTF(
     copy(child = newChild)
 }
 
+/**
+ * A logical plan that evaluates the 'analyze' method of a [[PythonUDTF]] on the executors.
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/pythonLogicalOperators.scala:278]: Maybe this is beyond the scope of this PR, but could we add more comments/examples here on how a polymorphic UDTF can be transformed into a plan with this node where it's `analyze` method can be evaluated on executors?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/AnalyzePythonUDTFOnExecutorsExec.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  override def evaluate(
+      argMetas: Array[ArgumentMetadata],
+      iter: Iterator[InternalRow],
+      schema: StructType,
+      context: TaskContext): Iterator[Iterator[InternalRow]] = {
+    val tableArgs = udtf.children.map(_.isInstanceOf[FunctionTableSubqueryArgumentExpression])
+    val parser = CatalystSqlParser
+    val runner = new UserDefinedPythonTableFunctionAnalyzeRunner(
```
Comment (allisonwang-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/AnalyzePythonUDTFOnExecutorsExec.scala:57]: This `UserDefinedPythonTableFunctionAnalyzeRunner` extends from the `PythonPlannerRunner`, which is designed to run Python processes on the driver side. Does it work out of the box on executors?  cc @ueshin

Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/AnalyzePythonUDTFOnExecutorsExec.scala:57]: As discussed offline, we need to make this a subclass of `BasePythonRunner`. I can look into that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala]: I reverted this part.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/AnalyzePythonUDTFOnExecutorsExec.scala (hunk 3 above)
Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/AnalyzePythonUDTFOnExecutorsExec.scala]: This is done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/pythonLogicalOperators.scala (hunk 2 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/pythonLogicalOperators.scala:278]: Sure, I added some more comments here to help explain.

**Review by dtenedor — COMMENTED** [review]
Thanks @allisonwang-db for your review! I responded to your comments, and also updated this PR to use "worker.py" per the suggestion from @ueshin. Please take another look.

**Comment by dtenedor** [comment]
cc @allisonwang-db @ueshin friendly ping, this is ready for review

**Comment by dtenedor** [comment]
I am going to close this since I am getting no response on the PR; will reopen later if we want to implement this feature.

**Pull request closed without merging**